File: minion/app/src/main/java/snakesoft/minion/Materia/CalendarServiceProxy.java
package snakesoft.minion.Materia;

import com.google.protobuf.InvalidProtocolBufferException;

import calendar.Calendar;
import common.Common;

/**
 * Created by snake on 11/24/17.
 */

/*
   rpc Query (TimeRange) returns (CalendarItems);
   rpc Next (NextQueryParameters) returns (CalendarItems);
   rpc DeleteItem (common.UniqueId) returns (common.OperationResultMessage);
   rpc EditItem (CalendarItem) returns (common.OperationResultMessage);
   rpc AddItem (CalendarItem) returns (common.UniqueId);
*/

public class CalendarServiceProxy
{
    public CalendarServiceProxy(MateriaConnection materiaConnection)
    {
        mMateriaConnection = materiaConnection;
    }

    public Calendar.CalendarItems query(Calendar.TimeRange input) throws InvalidProtocolBufferException, MateriaUnreachableException {
        return Calendar.CalendarItems.parseFrom(mMateriaConnection.sendMessage(
                input.toByteString(),
                "CalendarService",
                "Query"
        ));}

    public Calendar.CalendarItems next(Calendar.NextQueryParameters input) throws InvalidProtocolBufferException, MateriaUnreachableException {
        return Calendar.CalendarItems.parseFrom(mMateriaConnection.sendMessage(
                input.toByteString(),
                "CalendarService",
                "Next"
        ));}

    public void deleteItem(Common.UniqueId id) throws InvalidProtocolBufferException, MateriaUnreachableException {
        mMateriaConnection.sendMessage(
                id.toByteString(),
                "CalendarService",
                "DeleteItem"
        );
    }

    public Common.UniqueId addItem(Calendar.CalendarItem item) throws InvalidProtocolBufferException, MateriaUnreachableException {
        return Common.UniqueId.parseFrom(mMateriaConnection.sendMessage(
                item.toByteString(),
                "CalendarService",
                "AddItem"
        ));
    }

    public boolean editItem(Calendar.CalendarItem item) throws InvalidProtocolBufferException, MateriaUnreachableException {
        return Common.OperationResultMessage.parseFrom(mMateriaConnection.sendMessage(
                item.toByteString(),
                "CalendarService",
                "EditItem"
        )).getSuccess();
    }

    private MateriaConnection mMateriaConnection;
}

File: minion/app/src/main/java/snakesoft/minion/materia/CalendarServiceProxy.kt
package snakesoft.minion.materia

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import snakesoft.minion.Models.ITrackable
import snakesoft.minion.Models.StatusOfChange
import snakesoft.minion.Models.UUIDSerializer

@Serializable
data class CalendarItem(
        @Serializable(with = UUIDSerializer::class)
        override var id: java.util.UUID,
        var text: String,
        var timestamp: Long,
        var nodeReference: String = "",
        var typename: String = "calendar_item",
        var reccurencyTypeChoice: String = "None",
        var entityTypeChoice: String = "Task",
        var urgencyChoice: String = "Not Urgent",
        override var trackingInfo: StatusOfChange = StatusOfChange.None
) : ITrackable

@Serializable
data class CalendarParams(
        var text: String,
        var timestamp: Long,
        var reccurencyTypeChoice: String,
        var entityTypeChoice: String,
        var urgencyChoice: String
)

@Serializable
data class QueryResult(val id: String, val typename: String, val object_list: List<CalendarItem>)

@Serializable
data class CalendarModify(val operation: String = "modify", val id: String, val params: CalendarParams)

@Serializable
data class CalendarAdd(val operation: String, val typename: String, var params: CalendarParams)

class CalendarServiceProxy(private val mMateriaConnection: MateriaConnection) {

    @Throws(MateriaUnreachableException::class)
    fun query(): List<CalendarItem>
    {
        val jsonData = "{\"operation\": \"query\", \"filter\": \"IS(calendar_item)\"}"
        val resp = mMateriaConnection.sendMessage(jsonData)
        return format.decodeFromString(QueryResult.serializer(), resp).object_list
    }

    @Throws(MateriaUnreachableException::class)
    fun completeItem(id: java.util.UUID)
    {
        val jsonData = "{\"operation\": \"complete\", \"id\": \"${id}\"}"
        mMateriaConnection.sendMessage(jsonData)
    }

    @Throws(MateriaUnreachableException::class)
    fun addItem(item: CalendarItem)
    {
        val cp = CalendarParams(item.text, item.timestamp, item.reccurencyTypeChoice, item.entityTypeChoice, item.urgencyChoice)
        var ca = CalendarAdd("create", "calendar_item", cp)
        mMateriaConnection.sendMessage(Json.encodeToString(CalendarAdd.serializer(), ca))
    }

    @Throws(MateriaUnreachableException::class)
    fun editItem(item: CalendarItem): Boolean
    {
        val cp = CalendarParams(item.text, item.timestamp, item.reccurencyTypeChoice, item.entityTypeChoice, item.urgencyChoice)
        var ca = CalendarModify("modify", item.id.toString(), cp)
        mMateriaConnection.sendMessage(Json.encodeToString(CalendarModify.serializer(), ca))
        return true
    }
}
